File: #1261.find-elements-in-a-contaminated-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class FindElements {
public:
    FindElements(TreeNode* root) {
        
       root->val = 0;
       queue<pair<TreeNode*,int>> q;
       q.push({root, root->val});
       while (!q.empty()) {
            TreeNode *t=q.front().first;
            int v = q.front().second;
            m[v]++;
            q.pop();
            if (t->left) {
                q.push({t->left, 2 * v + 1});
            }

            if (t->right) {
                q.push({t->right, 2 * v + 2});
            }
       }
    }
    
    bool find(int target) {
        
        return m[target];
    }

private:
     map<int,int> m;
};

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements* obj = new FindElements(root);
 * bool param_1 = obj->find(target);
 */

File: #1261.find-elements-in-a-contaminated-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */
var FindElements = function(root) {
    this.set = new Set();
    root.val = 0;
    let queue = [root];
    while (queue.length) {
        let curr = queue.pop();
        this.set.add(curr.val);
        if (curr.left) {
            curr.left.val = curr.val * 2 + 1;
            queue.push(curr.left);
        }
        if (curr.right) {
            curr.right.val = curr.val * 2 + 2;
            queue.push(curr.right);
        }
    }
};

/** 
 * @param {number} target
 * @return {boolean}
 */
FindElements.prototype.find = function(target) {
    return this.set.has(target);
};

/** 
 * Your FindElements object will be instantiated and called as such:
 * var obj = new FindElements(root)
 * var param_1 = obj.find(target)
 */
